services/scheduler/scheduler.go

```diff
@@ -0,0 +1,327 @@
+
+	// Don't add it to scheduler, if it's disabled.
+	if dbTask.Disabled {
+		return nil
+	}
+
+	j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+	if !dbTask.StartAt.IsZero() {
+		j = j.StartAt(dbTask.StartAt)
+	}
+
+	fn := s.wrapTask(task, dbTask.ID)
+	if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	disabled := false
+	dbTasks, err := models.FindScheduledTasks(s.db.Querier, models.ScheduledTasksFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	tasks := make([]Task, 0, len(dbTasks))
+	for _, dbTask := range dbTasks {
+		task, err := s.convertDBTask(dbTask)
+		if err != nil {
+			return err
+		}
+		tasks = append(tasks, task)
+	}
+
+	s.scheduler.Clear()
+	for i, task := range tasks {
+		dbTask := dbTasks[i]
+		fn := s.wrapTask(task, dbTask.ID)
+		j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+		if !dbTask.StartAt.IsZero() {
+			j = j.StartAt(dbTask.StartAt)
+		}
+		if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+func (s *Service) wrapTask(task Task, id string) func() {
+	return func() {
+		var err error
+		l := s.l.WithFields(logrus.Fields{
+			"id":       id,
+			"taskType": task.Type(),
+		})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.taskMx.Lock()
+		s.tasks[id] = cancel
+		s.taskMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.taskMx.Lock()
+			delete(s.tasks, id)
+			s.taskMx.Unlock()
+		}()
+
+		t := time.Now()
+		l.Debug("Starting task")
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(true),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running state: %v", err)
+		}
+
+		taskErr := task.Run(ctx)
+		if taskErr != nil {
+			l.Error(taskErr)
+		}
+		l.WithField("duration", time.Since(t)).Debug("Ended task")
+
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(false),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running status: %v", err)
+		}
+
+		s.taskFinished(id, taskErr)
+	}
+}
+
+func (s *Service) taskFinished(id string, taskErr error) {
+	var job *gocron.Job
+	for _, j := range s.scheduler.Jobs() {
+		if len(j.Tags()) > 0 && j.Tags()[0] == id {
+			job = j
+			break
+		}
+	}
+	l := s.l.WithField("id", id)
+
+	dbTask, err := models.FindScheduledTaskByID(s.db.Querier, id)
+	if err != nil {
+		return
+	}
+
+	params := models.ChangeScheduledTaskParams{
+		Succeeded: pointer.ToUint(dbTask.Succeeded),
+		Failed:    pointer.ToUint(dbTask.Failed),
+		Running:   pointer.ToBool(false),
+	}
+
+	if taskErr == nil {
+		params.Succeeded = pointer.ToUint(dbTask.Succeeded + 1)
+		params.Error = pointer.ToString("")
+	} else {
+		params.Failed = pointer.ToUint(dbTask.Failed + 1)
+		params.Error = pointer.ToString(taskErr.Error())
```

Comment:
don't we store it in `job` or `artifact`?

Reply:
Yes. Error is in job. But scheduler doesn't necessary run only jobs. But I can remove it.